algebra/Atiyah.lean

```diff
@@ -0,0 +1,1695 @@
+
+theorem zero_mul : (zero_ideal α) * S₁ = zero_ideal α :=
+set.ext $ λ x,
+  ⟨λ hx, hx (zero_ideal α) $ λ z ⟨p, q, hp, hq, hz⟩, by simp [zero_ideal] at *; simp [hz, hp],
+   λ hx, by simp [zero_ideal] at *; rw hx; exact is_ideal.zero_mem _⟩
+
+theorem mul_assoc : (S₁ * S₂) * S₃ = S₁ * (S₂ * S₃) :=
+set.ext $ λ x,
+⟨λ hx, hx (S₁ * (S₂ * S₃)) $ λ z ⟨pq, r, hpq, hr, hz⟩,
+   let ⟨L, hL, hLpq⟩ := mem_span_of_mem_generate hpq in
+   have h : list.sum (list.map (λ b, b * r) L) ∈ S₁ * (S₂ * S₃) :=
+     λ S hs hss, @@is_ideal.sum_mem _ S hs (list.map (λ b, b * r) L) $ λ x hx,
+       let ⟨y, hyL, hyrx⟩ := list.exists_of_mem_map hx,
+           ⟨m, n, ⟨p, q, hp, hq, hmpq⟩, hmny⟩ := hL y hyL in
+       hss ⟨p, (q * r) * n,
+         hp,
+         λ T ht htt, @@is_ideal.mul_mem _ ht $ htt ⟨q, r, hq, hr, rfl⟩,
+         by rw [← hyrx, ← hmny, hmpq]; ac_refl⟩,
+   by rw [list.sum_mul, hLpq] at h; rwa hz,
+ λ hx, hx ((S₁ * S₂) * S₃) $ λ z ⟨p, qr, hp, hqr, hz⟩,
+   let ⟨L, hL, hLqr⟩ := mem_span_of_mem_generate hqr in
+   have h : list.sum (list.map (λ b, b * p) L) ∈ (S₁ * S₂) * S₃ :=
+     λ S hs hss, @@is_ideal.sum_mem _ S hs (list.map (λ b, b * p) L) $ λ x hx,
+       let ⟨y, hyL, hypx⟩ := list.exists_of_mem_map hx,
+           ⟨m, n, ⟨q, r, hq, hr, hmqr⟩, hmny⟩ := hL y hyL in
+       hss ⟨(p * q) * n, r,
+         λ T ht htt, @@is_ideal.mul_mem _ ht $ htt ⟨p, q, hp, hq, rfl⟩,
+         hr,
+         by rw [← hypx, ← hmny, hmqr]; ac_refl⟩,
+    by rw [list.sum_mul, hLqr, comm_ring.mul_comm] at h; rwa hz⟩
+
+theorem mul_add_subset : (S₁ * S₂) + (S₁ * S₃) ⊆ S₁ * (S₂ + S₃) :=
+λ x ⟨y, z, hy, hz, hx⟩,
+let ⟨Ly, hy, hLy⟩ := mem_span_of_mem_generate hy,
+    ⟨Lz, hz, hLz⟩ := mem_span_of_mem_generate hz in
+mem_generate_of_mem_span ⟨Ly ++ Lz,
+  λ p hp, or.cases_on (list.mem_append.1 hp)
+    (λ hpy, let ⟨py, pz, ⟨pyy, pyz, hpyy, hpyz, hpyyz⟩, hpyzp⟩ := hy p hpy in
+       ⟨py, pz, ⟨pyy, pyz, hpyy, subset_add_left hpyz, hpyyz⟩, hpyzp⟩)
+    (λ hpz, let ⟨py, pz, ⟨pzy, pzz, hpzy, hpzz, hpzyz⟩, hpyzp⟩ := hz p hpz in
+       ⟨py, pz, ⟨pzy, pzz, hpzy, subset_add_right hpzz, hpzyz⟩, hpyzp⟩),
+  by simp [list.sum_append, *]⟩
+
+theorem mul_add : S₁ * (S₂ + S₃) = (S₁ * S₂) + (S₁ * S₃) :=
+set.ext $ λ x,
+⟨λ hx, let ⟨L, hx, hLx⟩ := mem_span_of_mem_generate hx in
+   (list.rec_on L
+     (λ x hx hLx, by rw ← hLx; exact is_ideal.zero_mem _)
+     (λ hd tl ih x hx hLx, have h : hd + list.sum tl ∈ S₁ * S₂ + S₁ * S₃,
+       from is_ideal.add_mem
+         (let ⟨y, z, ⟨m, n, hm, ⟨ny, nz, hny, hnz, hnyz⟩, hymn⟩, hyz⟩ := hx hd (or.inl rfl) in
+            ⟨m * ny * z, m * nz * z,
+             is_ideal.mul_mem $ subset_generate _ ⟨m, ny, by simp [*]⟩,
+             is_ideal.mul_mem $ subset_generate _ ⟨m, nz, by simp [*]⟩,
+             by rw [← hyz, hymn, hnyz, mul_add, add_mul]⟩)
+         (ih (list.sum tl) (λ z hz, hx z (or.inr hz)) rfl),
+            by simp at hLx; rwa hLx at h))
+   x hx hLx,
+ λ hx, mul_add_subset hx⟩
+
```

Comment:
At this point you have probably proved that ideals form some absurd structure (a partially ordered semiring or something!). Mathematicians don't care about such structure, but if there's a class for it then you may as well make it an instance because you might then get some other theorems about ideals for free. Not sure if this is worth following up.